add key points to episode 3 (#322)
datacarpentry/python-ecology-lesson: _episodes/03-index-slice-subset.md

    - "Locate subsets of data using masks."
    - "Describe BOOLEAN objects in Python and manipulate data using BOOLEANs."
keypoints:

>     - "Data can be subset using column headings, numeric ranges, operators, or specific index locations."

Reviewer: ```suggestion
     - "In Python, portions of data can be accessed using indices, slices, column headings, and condition-based subsetting."
```

>     - "Data can be subset using column headings, numeric ranges , operators, or specific index locations."

Reviewer 2: No space before comma in `numeric ranges , `

Reviewer 2: another point: it's not super clear what `Data can be subset` means.

Reviewer: ```suggestion
    - "Data can be subset using column headings, numeric ranges, operators, or specific index locations."
```

Author: Very good point. Now I can't remember why I wanted to add this particular keypoint. Most of the lesson is about using column names to summarize data. Because really, this keypoint summaries this [quick guide to subsetting data in R](https://www.statmethods.net/management/subset.html).

I'm happy to remove it if you think it's irrelevant.

Author: space deleted in https://github.com/datacarpentry/python-ecology-lesson/pull/322/commits/034f56aaad5f0332b02d7d7cd1d38238ba8d1fc0

Reviewer: This seems to. have bled into the next point, so I'm going to mark this as resolved.

Reviewer 2: My point was that in this lesson we use column headings, ranges, etc for _accessing_ portions of data not for changing it. Because the verb `subset` means `be a subset of`, the phrase `Data can be subset` can be understood as `Data can be changed in-place`

Reviewer: Oh, interesting, I had not considered this interpretation. Is there a wording that might be better?

Reviewer 2: Nah :)

<br>
jk! I'm sure there is. Something like

>In Python, portions of data can be accessed using indices, slices, column headings (Panda's data frames), and condition-based subsetting (Panda's data frames and numpy arrays).

Author: LGTM

Author: This beta suggest and apply feature is awesome!

Reviewer 2: > Does this suggestion work for everyone?

Looks good as long as all of us understand what `condition-based subsetting` is :)

Reviewer: ```suggestion
    - "In Python, portions of data can be accessed using indices, slices, column headings, and condition-based subsetting."
```

I'll add it in - the beta suggestions thing is a little finnicky to if you changed a prior line.

>     - "Python uses 0-based indexing."

Reviewer 2: I suggest either adding a brief explanation or replacing the keypoint with it.
```
Python uses 0-based indexing: <brief explanation>
```
Explanation could be:
```
The first element in a list, tuple or any other data structure has an index of 0
```

Reviewer: ```suggestion
    - "Python uses 0-based indexing, in which the first element in a list, tuple or any other data structure has an index of 0."
```

Author: thanks april! suggestion incorporated in https://github.com/datacarpentry/python-ecology-lesson/pull/322/commits/034f56aaad5f0332b02d7d7cd1d38238ba8d1fc0

>     - "Data indexing, slicing and subsetting are common steps in data wrangling and exploration."

Reviewer 2: what message are you trying to convey with this keypoint? It seems too general and not Python-specific

Reviewer: I like that - this is the Python lesson, but tapping into a larger set of tasks and concerns.

Author: Thanks April. Yes, I was specifically trying to make a broader picture of why this is important. Maybe is could say  something like "The `pandas` library useful for data indexing, slicing and subsetting, which are common steps in data wrangling and exploration."

Reviewer 2: 1. I think keypoints should be specific to episodes
2. Indexing and slicing are not Pandas-specific.

Perhaps, you meant something like this
```
For data exploration and wrangling in Python you can use such powerful subsetting methods as indexing (accessing individual elements of complex data structures _via_ their index) and slicing (accessing several individual elements...)
```

Reviewer: I don't disagree, but the keypoint is specific to what we're doing in the lesson. If we want it to be more pythonic, I don't not like what you've written, but its a bit wordy. Perhaps

```suggestion
    - "Pandas enables common data exploration steps such as data indexing, slicing and subsetting."
```

Reviewer 2: Python native lists and tuples can be indexed and sliced too.
Pandas and numpy can do condition-based subsetting.

Reviewer: Sure, but I don't think it's on us to emphasize the uniqueness of the technology we teach.

Reviewer 2: `Pandas supports...` would make this keypoint more accurate.

On a different note: aren't indexing and slicing examples of subsetting? Or did we miss `conditional` before `subsetting`?

Reviewer 2: I think we ought to change line 37 to `condition-based` or `conditional` subsetting

Author: yep! i love this new beta feature

Reviewer: It really is great. It's exactly what I need for working with people who are new to LaTeX, too. Leaving line edits is always the biggest stumbling block for the new folks.
